Buff initial stacks

gastank opened the pull request:
Implement parsing spell data for initial stack of buffs, and applying that number of stacks by default when the buff is called via trigger() or execute() with no explicit stack value.

For reverse buffs, behavior remains unchanged where a default value call will results in decrement(1).
Similarly, decrement() and increment() remain unchanged as expected behavior when those are called with default value is that the buff is adjusted by a single stack each time.

* buff_t::set_initial_stack( int ) and buff_t::modify_initial_stack( int ) can be used to explicitly set/modify the initial stacks, overriding what's read off the spell data. Can be chained at buff construction like the *_max_stack() counterparts.
* buff_t::initial_stack() will return the _initial_stack value.

gastank committed:
[Buffs] implement _initial_stack variable & related functions

When a buff is trigger()'d or execute()'d with the default parameter of
`stack == -1` it will appropriately increment() a number of stacks equal
to the _intial_stack value (for reverse buffs, behavior is unchanged and
will call decrement(1)).

`buff_t::set_initial_stack( int )` to set the _initial_stack value.
`buff_t::modify_initial_stack( int )` to add to the current _initial_stack.

gastank committed:
[Druid] adjust buff triggers for initial_stack

Adjusted bloodtalons, eye of fearful symmetry, and warrior of elune

gastank committed:
Allow initial_stack to override max_stack

gastank committed:
Adjust for buff delay event combining proc triggers

nuoHep commented on engine/buff/sc_buff.hpp:
```diff
@@ -74,6 +74,7 @@ struct buff_t : private noncopyable
   const spell_data_t* trigger_data;
 
 public:
+  int _initial_stack;
```
Maybe move it adjacent to `_max_stack`?

gastank committed:
Use helper function to resolve default value parameter passed to buff_t::trigger()/execute()

scamille commented:
Can you maybe add some comments to the places where you introduced std::abs?

Without knowing the discussion in Discord, I would have a hard time knowing why you introduced those additions here.

I assume you added the abs adjustments to the trigger functions and not to the init function so that initial_stacks -1 is still a valid value in the buff, probably for reverse buffs?

gastank commented:
Yeah still working through. I don't think those abs() are even necessary. What do you mean by adjusting in the init function?

gastank committed:
check for default value(-1) instead of abs(), add more comments

gastank committed:
Remove duplicate check vs _max_stack

set_initial_stack checks & modifies _max_stack, so no need to do it again in modify_initial_stack

gastank committed:
Handle 'Add Flat Modifier (107): Spell Initial Stacks (4)' in buff_t::apply_affecting_aura

scamille commented:
What I meant is, why not ensure _initial_stacks is always > 0 already in the set_initial_stacks function instead of using abs functions in the trigger process. Unless of course negative values are perfectly valid, eg. for reverse buffs.

gastank commented:
That's already being done in`set_initial_stacks` to set `buff_t::_initial_stack` as 1 if the value in the spell data is negative.

The parts in trigger()/execute() is about handling the `stacks` parameter of those functions, which used to have default value of '1' but is being changed to '-1' to allow for calling `trigger(1)` to trigger a single stack instead of `_initial_stack`.

I am wondering however, since negative stacks in the spell data seems to indicate that the spell should have special/custom handling, it might be better to NOT guard against that in `set_initial_stack` and instead let it go through and assert out at execute(). So for those buffs, unless you:
1. explicitly set the initial stacks in buff construction, or
1. call trigger()/execute() with a stack value,

it will assert out to ensure that you do one of those things since the spell data indicates that the buff doesn't have a standard initial stack count.

scamille commented:
Sure why not. I assume the currently implemented buffs with -1 initial stack in spell data can be adjusted / already handle it, and you can add a little string literal inside the assert in trigger/execute explaining that this needs to be handled with an explicit stack value.

If it really only affects very few buffs and strongly correlates with "special handling required" and not just Blizzard having a mess in spell data, this sounds like a sensible idea.

gastank committed:
assert at execute() for buffs with negative stacks in the data

gastank committed:
don't allow -1 max_stacks

gastank committed:
assert in modify_initial_stack to alert when explicit set_initial_stack is required

gastank committed:
[Rogue] Explicitly set Flagellation initial stacks

gastank committed:
Revert "[Druid] adjust buff triggers for initial_stack"

This reverts commit ab9a178cf1211b0b39b25fc47bc85150be223b11.

gastank committed:
adjust assert messages
